parsers/UY.py

```diff
@@ -46,118 +62,182 @@ def get_salto_grande(session: Optional[Session]) -> float:
     return generation
 
 
-def parse_page(session: Optional[Session]):
-    r = session or Session()
-    url = "https://apps.ute.com.uy/SgePublico/ConsPotenciaGeneracionArbolXFuente.aspx"
-    response = r.get(url)
-    soup = BeautifulSoup(response.text, "html.parser")
+def parse_num(num_str: str) -> float:
+    """
+    given a string containing a decimal number in
+    south american format, e.g. '23.456,45' using ',' as a decimal
+    and '.' as a separator, convert to the corresponding float
+    :param num_str: string containing a number south american format,
+        e.g. '23.456,45' using ',' as a decimal
+    :returns: a float with the value denoted by num_str
+    """
+    return float(num_str.strip().replace(".", "").replace(",", "."))
+
+
+def correct_for_salto_grande(entry, session: Session):
+    """
+    corrects a single entry parsed from the UTE website using salto_grande data
+    Switched this to only operate on a single entry or list of entries so
+    fewer requests are needed
+    """
+    # https://github.com/tmrowco/electricitymap/issues/1325#issuecomment-380453296
+    salto_grande = get_salto_grande(session, entry["time"])
+    entry["hydro"] = entry["hydro"] + salto_grande
+    return entry
+
+
+# time:
+def parse_page(session: Session):
+    """
+    Queries the url in UTE_URL, and parses hourly production and trade data
+    and retruns the results as a list of dictionary objects
+    """
+    # load page
+    resp = session.get(UTE_URL)
+
+    # get the encoded XML with all the data we want
+    soup = BeautifulSoup(resp.text, "html.parser")
+    xml_tag = soup.find(id="valoresParaGraficar")
+    xml_string = f"<{xml_tag.contents[0]}>{xml_tag.contents[1]}"
```

Comment:
You need to check that the `xml_tag` is defined before calling it's properties, current behavior results in a type error.